content/posts/event-driven-scaling-fission-function-using-keda.md

```diff
@@ -0,0 +1,239 @@
+	producerConfig.Producer.RequiredAcks = sarama.WaitForAll
+	producerConfig.Producer.Retry.Max = 100
+	producerConfig.Producer.Retry.Backoff = 100
+	producerConfig.Producer.Return.Successes = true
+	producerConfig.Version = sarama.V1_0_0_0
+	producer, err := sarama.NewSyncProducer(brokers, producerConfig)
+	fmt.Println("Created a new producer ", producer)
+	if err != nil {
+		panic(err)
+	}
+	for msg := 1; msg <= 1000; msg++ {
+		ts := time.Now().Format(time.RFC3339)
+		message := fmt.Sprintf("{\"message_number\": %d, \"time_stamp\": \"%s\"}", msg, ts)
+		_, _, err = producer.SendMessage(&sarama.ProducerMessage{
+			Topic: "request-topic",
+			Value: sarama.StringEncoder(message),
+		})
+
+		if err != nil {
+			w.Write([]byte(fmt.Sprintf("Failed to publish message to topic %s: %v", "request-topic", err)))
+			return
+		}
+	}
+	w.Write([]byte("Successfully sent to request-topic"))
+}
+```
+
+Let’s create the environment and function
+```
+$ fission environment create --name go --image fission/go-env-1.12:1.10.0 --builder fission/go-builder-1.12:1.10.0
+$ fission fn create --name producer --env go --src producer.go --entrypoint Handler
+```
+
+### Creating Consumer Function
+
+The consumer function is a Node Js program. Save the following code with consumer.js
+
+```
+module.exports = async function (context) {
+    console.log(context.request.body);
+    let obj = context.request.body;
+    return {
+        status: 200,
+        body: "Consumer Response "+ JSON.stringify(obj)
+    };
+}
+```
+
+Let’s create the environment and function
+
+```
+$ fission env create --name nodeenv --image fission/node-env
+$ fission fn create --name consumer --env nodeenv --code consumer.js
+```
+
+## Creating Message Queue Trigger
+```
+$ fission mqt create --name kafkatest --function consumer --mqtype kafka --mqtkind keda --topic request-topic --resptopic response-topic --errortopic error-topic --maxretries 3 --metadata bootstrapServers=my-cluster-kafka-brokers.my-kafka-project.svc:9092 --metadata consumerGroup=my-group --metadata topic=request-topic  --cooldownperiod=30 --pollinginterval=5`
+```
+To take leverage advantage of KEDA scaling, one must set `--mqtkind keda`
+
+## Everything in Action
+
+Open one terminal and execute the following to keep watch on pods
+```
+$ kubectl get pods -w
+```
+Let’s subscribe to our response-topic to see the response return by consumer function invocations. Open another terminal and execute the following:
+```
+$ kubectl run kafka-consumer -ti --image=strimzi/kafka:latest-kafka-2.5.0 --rm=true --restart=Never -- bin/kafka-console-consumer.sh --bootstrap-server my-cluster-kafka-brokers.my-kafka-project.svc:9092 --topic response-topic
+```
+
+Since everything is in place, let’s invoke our producer function in another terminal
+```
+$ fission fn test --name producer
+```
+
+Viola! Now you can see eventually three new pods are being created in the first terminal and response of the consumer function in the second terminal. 
```

Comment:
Should we also show the Keda connector pod & HPA scaling from 0-1 here (May be you are showing in image, just in case not)

Reply:
Done